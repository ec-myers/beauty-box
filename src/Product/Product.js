import React from 'react';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import './Product.scss';

export const Product = ({ product, toggleCollection, collection }) => {
  // const { collection } = this.props;
  let tagList = product.tag_list.map(tag => <li>{tag}</li>);
  let { brand, name, description, product_link, id} = product;
  let isSaved = collection.map(product => product.id).includes(id) ? 'saved' : '';
  let btnMessage = isSaved ? 'REMOVE' : 'ADD';

  return (
    <div className={`Product ${isSaved}`}>
      <div className='image-div'>
        <object className='product-img' data={product.image_link} type='image/jpg' alt={name}>
          <img className='product-img' src={require('../Images/default_image.jpg')} alt='default' />
        </object>
      </div>
      <div className='border-div'></div>
      <h3>{brand}</h3>
      <h4>{name}</h4>
      <p>{description}</p>
      <ul className='tag-list'>{tagList}</ul>
      <footer>
        <div className='shop-div'>
          < a href={product_link}>SHOP</a>
        </div>
        <button onClick={() => toggleCollection(product)}>{btnMessage}</button>
      </footer>
    </div>
  )
}

export const mapStateToProps = ({collection}) => ({
  collection
});

export default connect(mapStateToProps)(Product);

Product.propTypes = {
  product: PropTypes.object,
  toggleCollection: PropTypes.func,
  collection: PropTypes.array
}