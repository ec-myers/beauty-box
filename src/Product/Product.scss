$background-color: rgb(255, 217, 218);
$font-color: rgb(5,29,65);
$button-hover-color: darken(rgb(191, 192, 192), 8%);

.Product {
  // cursor: pointer;
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.356);
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
  margin: 2% 1%;
  padding: 4px;
  transition: .25s;

  &:hover {
    box-shadow: 0 2px 10px #cc7178;
    transition: .25s;
  }

  .image-div {
    height: 285px;
    overflow: hidden;
    padding: 4px;
    width: 100%;

    .product-img {
      transform: translateY(-30px);
      width: 100%;
    }
  }
  
  .border-div {
    border: 2px solid $font-color;
    width: 100%;
  }

  h3 {
    font-size: 1.4em;
    margin: 15px;
  }

  h4 {
    margin-left: 15px;
  }

  p {
    height: 200px;
    margin: 4px;
    overflow: scroll;
    margin: 15px;
  }

  .tag-list-div {
    min-height: 25px;

    ul {
      display: flex;
      flex-direction: row;
      flex-wrap: wrap;
      justify-content: center;
      list-style-type: none;
      margin: 4px;
  
      li {
        background-color: $background-color;
        border-radius: 4px;
        font-size: 1em;
        margin: 6px;
      }
    }
 }

  footer {
    align-items: center;
    display: flex;
    flex-direction: row;
    height: 50px;
    justify-content: space-between;

    a {
      align-items: center;
      background-color: transparent;
      border-radius: 4px;
      border: 2px solid $font-color;
      color: $font-color;
      cursor: pointer;
      font-size: .8em;
      font-weight: bold;
      line-height: 32px;
      height: 35px;
      margin-left: 4px;
      text-align: center;
      text-decoration: none;
      width: 145px;
      
      &:hover {
        box-shadow: 0 1px 5px rgba(5, 29, 65, 0.774);
        color: royalblue;
        transition: 0.15s;        
      }
    }

    button {
      background-color: transparent;
      border: 2px solid $font-color;
       border-radius: 4px;
      color: $font-color;
      cursor: pointer;
      font-size: .8em;
      font-weight: bold;
      height: 35px;
      margin-right: 4px;
      width: 145px;

      &:hover {
        box-shadow: 0 1px 5px rgba(5, 29, 65, 0.774);
        color: royalblue;
        transition: 0.15s;      
      }
    } 
  }
}

.saved {
  background-color: lighten(rgb(255, 217, 218),5%);
}