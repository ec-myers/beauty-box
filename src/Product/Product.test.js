import React from 'react';
import { shallow } from 'enzyme';
import { Product, mapStateToProps} from './Product';

describe('Product', () => {
  let toggleCollection = jest.fn();
  let collection = [{}, {}];
  let product = {
    id: 1047,
    brand: "colourpop",
    name: "Blotted Lip",
    image_link: "https://file.com",
    product_link: "https://colourpop.com/",
    description: "Description",
    tag_list: ['tag1', 'tag2']
  }
  let wrapper = shallow(
  <Product 
    toggleCollection={toggleCollection}
    collection={collection}
    product={product}
  />);
  
  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call toggleCollection when button is clicked', () => {
    wrapper.find('button').simulate('click');
    expect(toggleCollection).toHaveBeenCalled();
  });
});

describe('mapStateToProps', () => {
  it('should return an array of objects', () => {
    let mockState = {
      collection: [{
      id: 1047,
      brand: "colourpop",
      name: "Blotted Lip",
      image_link: "https://file.com",
      product_link: "https://colourpop.com/",
      description: "Description",
      tag_list: ['tag1', 'tag2']
      }]
    };

    let expected = {
      collection: [{
      id: 1047,
      brand: "colourpop",
      name: "Blotted Lip",
      image_link: "https://file.com",
      product_link: "https://colourpop.com/",
      description: "Description",
      tag_list: ['tag1', 'tag2']
      }]
    };

    let mappedProps = mapStateToProps(mockState);

    expect(mappedProps).toEqual(expected);
  });
});