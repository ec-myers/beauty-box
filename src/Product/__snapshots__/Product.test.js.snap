// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product should match the snapshot 1`] = `
<div
  className="Product "
>
  <div
    className="image-div"
  >
    <object
      alt="Blotted Lip"
      className="product-img"
      data="https://file.com"
      type="image/jpg"
    >
      <img
        alt="default"
        className="product-img"
        src="default_image.jpg"
      />
    </object>
  </div>
  <div
    className="border-div"
  />
  <h3>
    colourpop
  </h3>
  <h4>
    Blotted Lip
  </h4>
  <p>
    Description
  </p>
  <div
    className="tag-list-div"
  >
    <ul
      className="tag-list"
    >
      <li>
        tag1
      </li>
      <li>
        tag2
      </li>
    </ul>
  </div>
  <footer>
    <a
      href="https://colourpop.com/"
    >
      SHOP
    </a>
    <button
      onClick={[Function]}
    >
      ADD
    </button>
  </footer>
</div>
`;
